### ISSUE TITLE
SDXL LoRA Training Stalls at Step 0 (0%) after All Setup in Clean Environment (PyTorch 2.3.0+cu121)

### ISSUE THREAD

**Issue by Jjungbong** [kijai/ComfyUI-FluxTrainer#168]
Hello Kijai and FluxTrainer Users,

I'm encountering an issue where SDXL LoRA training using ComfyUI-FluxTrainer (LATEST version from ComfyUI Manager, confirmed version 1.0.2 from workflow JSON `properties.ver`) stalls at the very beginning of the training steps (0% progress). This happens even after a complete clean install of Python, ComfyUI, all dependencies (including PyTorch 2.3.0+cu121), and using the official `sdxl_train_example_01.json` workflow with minimal path modifications for a single concept.

## Environment:
* **ComfyUI Version:** v0.3.39 (as shown in startup log)
* **Python Version:** 3.11.8 (installed fresh in `C:\Python_ComfyNew\`)
* **PyTorch Version:** 2.3.0+cu121 (installed via pip with official CUDA 12.1 index into the new Python environment)
* **xformers Version:** 0.0.26.post1 (installed via pip into the new Python environment)
* **ComfyUI-FluxTrainer Version:** LATEST (installed via ComfyUI Manager in a clean ComfyUI setup, workflow JSON shows `ver: "1.0.2"`)
* **OS:** Windows (Version 10.0.26100.4061 from terminal log)
* **GPU:** NVIDIA GeForce RTX 4090 (24GB VRAM)

## Steps Taken (Clean Install Process and Test):
1.  Uninstalled all previous Python versions. System rebooted.
2.  Installed Python 3.11.8 fresh into `C:\Python_ComfyNew\` (NOT added to system PATH).
3.  Installed PyTorch 2.3.0+cu121, torchvision 0.18.0+cu121, torchaudio 2.3.0+cu121 into the new Python environment.
4.  Installed xformers 0.0.26.post1 into the new Python environment.
5.  Installed ComfyUI fresh into `C:\ComfyUI_Clean\` using Git.
6.  Ran `C:\Python_ComfyNew\python.exe -m pip install -r requirements.txt` in the `C:\ComfyUI_Clean\` directory (All ComfyUI dependencies installed successfully).
7.  Installed ComfyUI Manager in the new ComfyUI.
8.  Installed `Kijai/ComfyUI-FluxTrainer` (LATEST) and all its prompted dependencies (including other custom nodes required by the example workflow like `rgthree-comfy`, `ComfyUI-LogicUtils`, `comfyui-kjnodes`) via Manager.
9.  Restarted ComfyUI. All custom nodes, including FluxTrainer, load without Python import errors in the startup log.

## Problem Description (Using Kijai's `sdxl_train_example_01.json`):
1.  Loaded the `sdxl_train_example_01.json` workflow into the clean ComfyUI.
2.  Modified the example workflow *only* for a single concept test:
    * **SDXLModelSelect node:** `checkpoint` set to local `sd_xl_base_1.0.safetensors` (located in `C:\ComfyUI_Clean\models\checkpoints\`).
    * **TrainDatasetAdd node:**
        * `dataset_path`: Set to single concept folder: `C:\ComfyUI_Clean\datasets\lora_training\10_dokingi_mustache` (This folder contains 17 images and corresponding .txt caption files. Dataset files were copied to this new clean path.)
        * `class_tokens`: Set to `dokingi_mustache`.
        * `num_repeats`: Set to `10`.
    * **TrainDatasetGeneralConfig node:** `shuffle_caption: True`, `caption_extension: .txt` (no spaces).
    * **InitSDXLLoRATraining node:**
        * `output_name`: `dokingi_clean_final_test`
        * `output_dir`: `models/loras`
        * `network_dim`: `32` (User preferred, example default was 16)
        * `network_alpha`: `32` (User preferred, example default was 16)
        * `learning_rate`: `1.0` (Kept example's Prodigy optimizer default LR)
        * `max_train_steps`: `750` (Matching example's first loop segment)
        * `cache_text_encoder_outputs`: `disabled` (To avoid conflict with `shuffle_caption: True`)
        * `train_text_encoder`: `clip_l`
        * `clip_l_lr`: `5e-6`
        * `clip_g_lr`: `0.0`
        * Other parameters (optimizer, etc.) kept as per the example workflow.
    * **FluxTrainLoop node:** `steps` set to `750`.
    * Simplified the example to only run the first training loop and save stage.
    * Workflow saved.
3.  When "Queue Prompt" (or "Run") is clicked, the ComfyUI terminal output shows:
    * `sd-scripts` (`train_network.py`) starts.
    * Dataset config is loaded correctly, showing `image_dir` pointing to the single concept folder (`...\\10_dokingi_mustache`), `class_tokens` as `dokingi_mustache`, and `num_repeats` as `10`.
    * It correctly finds `17 image files` and calculates `170 train images with repeating`.
    * All models (U-Net, Text Encoders, VAE) load successfully ("All keys matched successfully").
    * LoRA network is created and applied.
    * Optimizer (Prodigy) is prepared.
    * Training parameters (num_epochs: 5, total_optimization_steps: 750) are calculated correctly.
    * The `running training` message appears.
    * The `tqdm` progress bar `steps:   0%|          | 0/750 [00:00<?, ?it/s]` is printed.
    * Then, ComfyUI's `INFO Prompt executed in X.XX seconds` message appears.
    * **After this, there are NO further updates in the terminal/log. The training remains stuck at `steps: 0%`, and no loss values or step increments are printed, even after waiting for several minutes.** GPU utilization is also low after an initial spike, indicating no active training computation.

## Log Snippet (From the latest clean environment test, showing stall at 0%):
```text
got prompt
queue_counter: 0
additional_args: --no_half_vae 
2025-06-01 23:31:17 INFO     Loading dataset config from   train_network.py:342
                               [[datasets]]                                   
                               resolution = [ 1024, 1024,]                    
                               batch_size = 1                                 
                               enable_bucket = true                           
                               bucket_no_upscale = false                      
                               min_bucket_reso = 256                          
                               max_bucket_reso = 1024                         
                               [[datasets.subsets]]                           
                               image_dir =                                    
                               "C:\\ComfyUI_Clean\\datasets\\                 
                               lora_training\\10_dokingi_mu                   
                               stache"                                        
                               class_tokens =                                 
                               "dokingi_mustache"                             
                               num_repeats = 10                               
                                                                              
                               [general]                                      
                               shuffle_caption = true                         
                               caption_extension = ".txt"                     
                               keep_tokens_separator = "|||"                  
                               caption_dropout_rate = 0.0                     
                               color_aug = false                              
                               flip_aug = false                               
                     INFO     prepare images.                       train_util.py:1955
                     INFO     get image size from name of           train_util.py:1864
                               cache files                                    

  0%|          | 0/17 [00:00<?, ?it/s]
100%|██████████| 17/17 [00:00<?, ?it/s]
                     INFO     set image size from cache             train_util.py:1894
                               files: 17/17                                   
                     INFO     found directory                       train_util.py:1896
                               C:\ComfyUI_Clean\datasets\lora_      
                               training\10_dokingi_mustache           
                               contains 17 image files                

read caption:   0%|          | 0/17 [00:00<?, ?it/s]
read caption: 100%|██████████| 17/17 [00:00<00:00, 330.81it/s]
2025-06-01 23:31:18 INFO     Found captions for 17 images.     train_util.py:1926
                     INFO     170 train images with             train_util.py:1996
                               repeating.                                     
                     INFO     0 reg images.                         train_util.py:1999
                     WARNING  no regularization images /          train_util.py:2004
                               正則化画像が見つかりませんでした                         
                     INFO     [Dataset 0]                           config_util.py:567
                               batch_size: 1                                  
                               resolution: (1024, 1024)                       
                               enable_bucket: True                            
                               network_multiplier: 1.0                        
                               min_bucket_reso: 256                           
                               max_bucket_reso: 1024                          
                               bucket_reso_steps: 64                          
                               bucket_no_upscale: False                       
                               [Subset 0 of Dataset 0]                        
                                 image_dir:                                   
                               "C:\ComfyUI_Clean\datasets\lora  
                               _training\10_dokingi_mustache"                 
                                 image_count: 17                             
                                 num_repeats: 10                             
                                 shuffle_caption: True                        
                                 keep_tokens: 0                               
                                 keep_tokens_separator: |||                   
                                 caption_separator: ,                         
                                 secondary_separator: None                    
                                 enable_wildcard: False                       
                                 caption_dropout_rate: 0.0                    
                                 caption_dropout_every_n_epo                  
                               chs: 0                                         
                                 caption_tag_dropout_rate:                    
                               0.0                                            
                                 caption_prefix: None                         
                                 caption_suffix: None                         
                                 color_aug: False                             
                                 flip_aug: False                              
                                 face_crop_aug_range: None                    
                                 random_crop: False                           
                                 token_warmup_min: 1                          
                                 token_warmup_step: 0.0                       
                                 alpha_mask: False                            
                                 custom_attributes: {}                         
                                 is_reg: False                                
                                 class_tokens:                                 
                               dokingi_mustache                               
                                 caption_extension: .txt                      
                     INFO     [Dataset 0]                           config_util.py:573
                     INFO     loading image sizes.                  train_util.py:918

  0%|          | 0/17 [00:00<?, ?it/s]
100%|██████████| 17/17 [00:00<?, ?it/s]
                     INFO     make buckets                          train_util.py:924
                     INFO     number of images (including       train_util.py:970
                               repeats) /                                     
                               各bucketの画像枚数（繰り返し回数を含む）                      
                     INFO     bucket 0: resolution (1024,       train_util.py:975
                               1024), count: 170                              
                     INFO     mean ar error (without repeats):     train_util.py:980
                               0.0                                            
                     INFO     preparing accelerator                 train_network.py:407
accelerator device: cuda
                     INFO     loading model for process         sdxl_train_util.py:33
                               0/1                                              
                     INFO     load StableDiffusion              sdxl_train_util.py:74
                               checkpoint:                                    
                               C:\ComfyUI_Clean\models\chec                   
                               kpoints\sd_xl_base_1.0.safet                   
                               ensors                                         
                     INFO     building U-Net                      sdxl_model_util.py:198
                     INFO     loading U-Net from                  sdxl_model_util.py:202
                               checkpoint                                     
2025-06-01 23:31:19 INFO     U-Net: <All keys matched             sdxl_model_util.py:208
                               successfully>                                  
                     INFO     building text encoders              sdxl_model_util.py:211
2025-06-01 23:31:20 INFO     loading text encoders from          sdxl_model_util.py:264
                               checkpoint                                     
                     INFO     text encoder 1: <All keys          sdxl_model_util.py:278
                               matched successfully>                          
                     INFO     text encoder 2: <All keys          sdxl_model_util.py:282
                               matched successfully>                          
                     INFO     building VAE                          sdxl_model_util.py:285
                     INFO     loading VAE from checkpoint         sdxl_model_util.py:290
                     INFO     VAE: <All keys matched               sdxl_model_util.py:293
                               successfully>                                  
2025-06-01 23:31:21 INFO     Enable memory efficient               train_util.py:3213
                               attention for U-Net                            
import network module: .networks.lora
                     INFO     [Dataset 0]                           train_util.py:2479
                     INFO     caching latents with caching          train_util.py:1026
                               strategy.                                      
                     INFO     caching latents...                    train_util.py:1075

  0%|          | 0/17 [00:00<?, ?it/s]
100%|██████████| 17/17 [00:00<00:00, 169.92it/s]
100%|██████████| 17/17 [00:00<00:00, 169.92it/s]
2025-06-01 23:31:22 INFO     create LoRA network. base dim (rank):  lora.py:928
                               32, alpha: 32.00000000000001                   
                     INFO     neuron dropout: p=None, rank dropout:  lora.py:929
                               p=None, module dropout: p=None                 
                     INFO     create LoRA for Text Encoder 1:       lora.py:1020
                     INFO     create LoRA for Text Encoder 2:       lora.py:1020
                     INFO     create LoRA for Text Encoder: 264      lora.py:1028
                               modules.                                       
                     INFO     create LoRA for U-Net: 722 modules.   lora.py:1036
                     INFO     enable LoRA for text encoder: 264      lora.py:1077
                               modules                                        
                     INFO     enable LoRA for U-Net: 722 modules    lora.py:1082
prepare optimizer, data loader etc.
                     INFO     use ProdigyScheduleFree               train_util.py:4863
                               optimizer |                                    
                               {'use_bias_correction': True}                  
running training
 num train images * repeats: 170
 num reg images: 0
 num batches per epoch: 170
 num epochs: 5 
 batch size per device: 1
 gradient accumulation steps: 1
 total optimization steps: 750 
2025-06-01 23:31:28 INFO     unet dtype: torch.bfloat16,         train_network.py:1124
                               device: cuda:0                                 
                     INFO     text_encoder [0] dtype:           train_network.py:1130
                               torch.bfloat16, device:                         
                               cuda:0                                         
                     INFO     text_encoder [1] dtype:           train_network.py:1130
                               torch.bfloat16, device:                         
                               cuda:0                                         

steps:   0%|          | 0/750 [00:00<?, ?it/s]
2025-06-01 23:31:29 INFO     Prompt executed in 12.40 seconds        main.py:189 
# (And no further log output for several minutes after this, training stalls at step 0)


### Questions for the Developer/Community:

- Is this "stall at step 0%" (after all preparations complete successfully, including correct data loading) a known issue with ComfyUI-FluxTrainer when using PyTorch 2.3.0+cu121 (or PyTorch 2.2.0+cu121, as it also stalled there with correctly configured data)?

- The FluxTrainer "Deep Research" report (and some community posts) "strongly recommended" PyTorch 2.4.0+. Could using PyTorch 2.2.0 or 2.3.0 cause this kind of silent stall at the beginning of the training loop? Are there specific features or behaviors in the training loop of the bundled sd-scripts (or FluxTrainer's handling of it) that absolutely require PyTorch 2.4+ for SDXL LoRA training to proceed past step 0?

- Are there any specific accelerate configurations (default config used) or xformers (v0.0.26.post1) interactions known to cause such a stall with SDXL training in FluxTrainer?

- What could be other reasons for the training to stall at step 0 with no error messages after all data and models are loaded and the running training message appears?

- Is there any specific additional_args for InitSDXLLoRATraining or other node settings that might help diagnose or resolve this "stall at step 0%"?

Any insights or suggestions would be greatly appreciated. We have performed a full clean install of Python, ComfyUI, and all dependencies, and used the example workflow with minimal path changes, but the training still stalls right after printing the initial 0% progress bar.

Thank you!
